crates/paralegal-policy/src/context.rs

```diff
@@ -219,38 +320,85 @@ impl Context {
             .flat_map(|v| v.iter())
     }
 
-    /// Returns an iterator over all sinks marked with `marker` out of the provided `dsts`.
-    pub fn marked_sinks<'a>(
-        &'a self,
-        dsts: impl IntoIterator<Item = &'a DataSink> + 'a,
-        marker: Marker,
-    ) -> impl Iterator<Item = &'a DataSink> + 'a {
-        dsts.into_iter().filter(move |dst| match dst {
-            DataSink::Argument { function, arg_slot } => self
+    /// Returns whether the given Node has the marker applied to it directly or via its type.
+    pub fn has_marker(&self, marker: Marker, node: Node) -> bool {
+        // TODO: check the type as well
+        match node.node {
+            NodeType::CtrlArgument(source) => match source {
+                DataSource::Argument(arg) => self
+                    .marker_to_ids
+                    .get(&marker)
+                    .map(|markers| {
+                        markers.iter().any(|(id, refinement)| {
+                            id == &node.ctrl_id
+                                && (refinement.on_self()
+                                    || refinement.on_argument().contains(*arg as u32).unwrap())
+                        })
+                    })
+                    .unwrap_or(false),
+                DataSource::FunctionCall(_) => {
+                    panic!("impossible - CtrlArgument variant cannot be DataSource::FunctionCall")
+                }
+            },
+            NodeType::CallSite(cs) => self
                 .marker_to_ids
                 .get(&marker)
                 .map(|markers| {
                     markers.iter().any(|(id, refinement)| {
-                        id == &function.function
-                            && (refinement.on_self()
-                                || refinement.on_argument().contains(*arg_slot as u32).unwrap())
+                        id == &cs.function && (refinement.on_self() || refinement.on_return())
                     })
                 })
                 .unwrap_or(false),
-            _ => false,
-        })
+            NodeType::CallArgument(ds) => match ds {
+                DataSink::Argument { function, arg_slot } => self
```

Comment:
The body of this branch seems mostly like a duplicate of `CtrlArgument`? Maybe we can refactor this logic into either a separate function or just a closure that we call from both?

Reply:
done